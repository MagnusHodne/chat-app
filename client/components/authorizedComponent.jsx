import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoadingComponent } from "./feedbackComponents";
import { ChatApiContext } from "../chatApiContext";

/**
 * Simple wrapper for components that require the user to be authorized
 * @param user the user object used for checking verification
 * @param {boolean} writeAccess Whether this component needs write access (default is false)
 * @param successComponent
 * @param failComponent
 * @param verificationFunction what function to use for verifying user. By default it uses the one specified in apiMethods
 */
export function AuthorizedComponent({
  user,
  writeAccess = false,
  successComponent = <div>Success</div>,
  failComponent = <div>Fail</div>,
}) {
  const navigate = useNavigate();
  const [verified, setVerified] = useState(false);
  const { verifyUser } = useContext(ChatApiContext);
  useEffect(() => {
    if (!verifyUser(user)) {
      navigate("/startlogin");
    } else {
      setVerified(true);
    }
  }, []);
  if (!verified)
    return <LoadingComponent message={"Redirecting to login..."} />;
  return successComponent;
}
