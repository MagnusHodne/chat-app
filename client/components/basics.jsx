import React from "react";
import { Link, useNavigate } from "react-router-dom";

export function FAIcon({ icon, className }) {
  return <i className={`icon ${icon} ${className}`}></i>;
}

export function Button({ title, icon, to, onClick, className }) {
  if (to) {
    const navigate = useNavigate();
    onClick = () => {
      navigate(to);
    };
  }
  const button = () => (
    <button
      className={`button ${className !== undefined ? className : ""}`}
      onClick={onClick}
    >
      {icon && <FAIcon icon={icon} />}
      {title}
    </button>
  );
  return button();
}
