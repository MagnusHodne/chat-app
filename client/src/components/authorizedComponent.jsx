import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoadingComponent } from "./feedbackComponents";
import { ChatApiContext } from "../chatApiContext";
import { UserContext } from "../userContext";

/**
 * Simple wrapper for components that require the user to be authorized
 * @param user the user object used for checking verification
 * @param {boolean} writeAccess Whether this component needs write access (default is false)
 * @param successComponent
 * @param failComponent
 */
export function AuthorizedComponent({
  writeAccess = false,
  successComponent = <div>Success</div>,
  failComponent = <div>Fail</div>,
}) {
  const navigate = useNavigate();
  const [verified, setVerified] = useState(false);
  const { user } = useContext(UserContext);
  const { verifyUser } = useContext(ChatApiContext);
  useEffect(() => {
    if (!verifyUser(user)) {
      navigate("/startlogin");
    } else {
      setVerified(true);
    }
  }, []);
  if (!verified)
    return <LoadingComponent message={"Redirecting to login..."} />;
  return successComponent;
}
