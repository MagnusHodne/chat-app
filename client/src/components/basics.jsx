import React from "react";
import { Link, useNavigate } from "react-router-dom";

export function FAIcon({ icon, className }) {
  return <i className={`${icon} ${className}`}></i>;
}

export function Button({ title, icon, to, onClick }) {
  if (to) {
    const navigate = useNavigate();
    onClick = () => {
      navigate(to);
    };
  }
  const button = () => (
    <button
      className={`p-2 m-1 rounded-md gap-2 text-thischord-100 bg-thischord-500 flex items-center justify-center hover:bg-thischord-400`}
      onClick={onClick}
    >
      {icon && <i className={`shrink-0 text-thischord-100 ${icon}`} />}
      {title}
    </button>
  );
  return button();
}
