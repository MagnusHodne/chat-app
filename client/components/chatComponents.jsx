import React, { useState } from "react";
import { FAIcon } from "./basics";

export function ChatHeader({ name }) {
  return (
    <div className={"chat-header"}>
      <FAIcon icon={"fa-solid fa-hashtag"} />
      <h3>{name}</h3>
    </div>
  );
}

function ChatMessageActions({ user }) {}

function ChatMessage({ message, user, created, _id, handleDelete }) {
  const date = new Date(created);

  return (
    <div className={"message-container"}>
      <div className={"message-info"}>
        <strong>{user.name}</strong>
        <small className={"message-time"}>
          {`${date.toDateString()} ${date.getHours()}:${date.getMinutes()}`}
        </small>
      </div>
      {message}
      <button className={"message-actions"} onClick={() => handleDelete(_id)}>
        Delete
      </button>
    </div>
  );
}

/**
 *
 * @param messages List of messages
 * @param onNewMessage Function for handling saves
 * @param onDeleteMessage Function for handling deletes
 * @param chatRoom Name of the chat room
 * @returns {JSX.Element}
 * @constructor
 */
export function ChatComponent({
  messages,
  onNewMessage,
  onDeleteMessage,
  chatRoom = "main",
}) {
  const [message, setMessage] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }

  return (
    <div className={"chat-window"}>
      <ChatHeader name={chatRoom} />
      <div className={"scroll"}>
        {messages.map(({ message, user, created, _id }) => (
          <ChatMessage
            message={message}
            user={user}
            created={created}
            key={_id}
            _id={_id}
            handleDelete={onDeleteMessage}
          />
        ))}
      </div>
      <footer>
        <form onSubmit={handleSubmit}>
          <input
            autoFocus={true}
            placeholder={`Message #${chatRoom}`}
            value={message}
            onChange={(e) => {
              setMessage(e.target.value);
            }}
          />
          <button>Send</button>
        </form>
      </footer>
    </div>
  );
}
