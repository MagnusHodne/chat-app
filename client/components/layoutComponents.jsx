import React, { useContext } from "react";
import { Button, FAIcon } from "./basics";
import { ChatApiContext } from "../chatApiContext";
import { UserContext } from "../userContext";

export function Content({ content }) {
  return <div className={"content"}>{content}</div>;
}

function UserActions() {
  const { user } = useContext(UserContext);
  const { verifyUser } = useContext(ChatApiContext);
  if (!verifyUser(user)) {
    return <></>;
  }

  return (
    <>
      <Button to={"/login/endsession"} title={"Log out"} className={"danger"} />
    </>
  );
}
export function Header() {
  return (
    <header className={"header"}>
      <div className={"title"}>
        <FAIcon icon={"fa-solid fa-gamepad"} />
        <h1>Thischord</h1>
      </div>
      <div>
        <UserActions />
      </div>
    </header>
  );
}

export function Sidebar({ content }) {
  return <div className={"sidebar"}>{content}</div>;
}
