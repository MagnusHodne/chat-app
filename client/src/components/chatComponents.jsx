import React, { useContext, useState } from "react";
import { Button, FAIcon } from "./basics";
import { UserContext } from "../userContext";

export function ChatHeader({ name }) {
  return (
    <div className={"flex flex-row items-center gap-2 pl-2"}>
      <FAIcon icon={"fa-solid fa-hashtag"} />
      <h3 className={"font-black"}>{name}</h3>
    </div>
  );
}

function ChatMessageAction({ icon, onClick, className }) {
  return (
    <button
      className={`border-none px-2 py-1 hover:bg-thischord-500 ${className}`}
      onClick={() => onClick()}
    >
      <FAIcon icon={icon} />
    </button>
  );
}

function ChatMessageActions({ author, handleDelete }) {
  const { user } = useContext(UserContext);
  const userIsAuthor = author.sub === user.sub;
  return (
    <div
      className={
        "absolute right-2 top-0 -translate-y-1/2 flex-row items-center justify-center self-end overflow-clip rounded border border-solid border-thischord-800 bg-thischord-600"
      }
    >
      <ChatMessageAction
        icon={"fa-solid fa-heart"}
        onClick={() => alert("Not yet implemented")}
      />

      {userIsAuthor && (
        <>
          <ChatMessageAction
            icon={"fa-solid fa-pencil"}
            onClick={() => alert("Not yet implemented")}
          />
          <ChatMessageAction
            icon={"fa-solid fa-trash-can"}
            onClick={() => handleDelete()}
            className={"text-red-600 hover:bg-red-600 hover:text-white"}
          />
        </>
      )}
    </div>
  );
}

function ChatMessage({ message, info, onDeleteMessage, displayInfo }) {
  const created = new Date(info.created);
  let date;
  if (Date.now() - created.getTime() < 1000 * 3600 * 24) {
    date = "Today";
  } else if (Date.now() - created.getTime() < 1000 * 3600 * 48) {
    date = "Yesterday";
  } else {
    date = created.toDateString();
  }
  const hours =
    created.getHours() < 10 ? `0${created.getHours()}` : created.getHours();
  const minutes =
    created.getMinutes() < 10
      ? `0${created.getMinutes()}`
      : created.getMinutes();
  const [showActions, setShowActions] = useState(false);

  return (
    <div
      className={"relative flex flex-col hover:bg-thischord-700"}
      onMouseEnter={() => setShowActions(true)}
      onMouseLeave={() => setShowActions(false)}
    >
      {showActions && (
        <ChatMessageActions
          author={info.user}
          handleDelete={() => onDeleteMessage(info._id)}
        />
      )}
      {displayInfo && (
        <div className={"mt-2 flex gap-2 pb-0"}>
          <div className={"inline-block"}>
            <span className={"font-extrabold"}>{info.user.name}</span>
          </div>

          <div className={"inline-block"}>
            <span className={"text-xs font-light text-[#99aab5]"}>
              {`${date} ${hours}:${minutes}`}
            </span>
          </div>
        </div>
      )}
      <div className={"py-0.5 px-0"}>{message}</div>
    </div>
  );
}

export function ChatComponent({
  messages,
  onNewMessage,
  onDeleteMessage,
  chatRoom = "main",
}) {
  const [message, setMessage] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }

  return (
    <div className={"grid h-full grid-rows-[3em_1fr_auto]"}>
      <ChatHeader name={chatRoom} />
      <div className={"scrollbar overflow-y-auto overflow-x-hidden pt-3"}>
        {messages.map(({ message, user, created, _id }, index) => {
          let displayInfo = false;
          if (index === 0) {
            displayInfo = true;
          } else if (messages[index - 1].user.sub !== user.sub) {
            displayInfo = true;
          } else if (
            new Date(created).getTime() -
              new Date(messages[index - 1].created).getTime() >
            1000 * 3600
          ) {
            displayInfo = true;
          }

          return (
            <ChatMessage
              message={message}
              info={{ user, created, _id }}
              displayInfo={displayInfo}
              key={_id}
              onDeleteMessage={onDeleteMessage}
            />
          );
        })}
      </div>
      <footer>
        <form
          className={"m-0 flex h-full flex-row items-stretch gap-1"}
          onSubmit={handleSubmit}
        >
          <input
            className={
              "flex-1 rounded border-none bg-thischord-500 px-3 py-0 placeholder:text-thischord-300"
            }
            autoFocus={true}
            placeholder={`Message #${chatRoom}`}
            value={message}
            onChange={(e) => {
              setMessage(e.target.value);
            }}
          />
          <Button title={"Send"} className={"m-0 h-full"} />
        </form>
      </footer>
    </div>
  );
}
