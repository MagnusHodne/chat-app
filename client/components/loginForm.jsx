import React from "react";
import { Button } from "./basics";
import { Header } from "./layoutComponents";

export function LoginForm() {
  return (
    <div className={"page-fullsize"}>
      <Header />
      <div className={"content"}>
        <div className={"login-buttons"}>
          <h1>Welcome!</h1>
          <p>Please log in</p>
          <div
            style={{ display: "flex", flexDirection: "column", width: "15em" }}
          >
            <Button
              to={"/login/google"}
              title={"Log in with Google"}
              icon={"fa-brands fa-google"}
            />
            <Button
              to={"/login/azure"}
              title={"Log in with Active Directory"}
              icon={"fa-brands fa-microsoft"}
            />
          </div>
        </div>
      </div>
    </div>
  );
}
