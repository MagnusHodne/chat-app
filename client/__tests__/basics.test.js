import React from "react";
import ReactDOM from "react-dom";
import { Button } from "../src/components/basics";
import { Simulate } from "react-dom/test-utils";

let element;
beforeEach(() => {
  element = document.createElement("div");
});

describe("Button", () => {
  it("matches snapshot", () => {
    ReactDOM.render(
      <Button title={"Test button"} icon={"fa-solid fa-comment"} />,
      element
    );
    expect(element).toMatchSnapshot();
  });
  it("calls the onclick function", () => {
    const onClick = jest.fn();

    ReactDOM.render(
      <Button
        title={"Test button"}
        icon={"fa-solid fa-comment"}
        onClick={onClick}
      />,
      element
    );
    Simulate.click(element.querySelector("button"));
    expect(onClick).toHaveBeenCalled();
  });
});
