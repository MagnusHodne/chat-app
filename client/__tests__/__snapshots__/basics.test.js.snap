// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<div>
  <button
    class="m-1 flex items-center justify-center gap-2 rounded-md bg-thischord-500 px-3 py-2 text-thischord-100 hover:bg-thischord-400 undefined"
  >
    <i
      class="shrink-0 text-thischord-100 fa-solid fa-comment"
    />
    Test button
  </button>
</div>
`;
